## 🛠️ 4. Import Libraries
We import all necessary numerical, data manipulation, plotting, and statistical libraries.


In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go
import plotly.io as pio
import warnings
warnings.filterwarnings('ignore')

# Set display settings for Pandas
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', lambda x: '%.4f' % x)

# Matplotlib & Seaborn Styling
sns.set_theme(style='whitegrid')
plt.rcParams['figure.figsize'] = (10, 6)
plt.rcParams['axes.titlesize'] = 16
plt.rcParams['axes.labelsize'] = 12
plt.rcParams['xtick.labelsize'] = 11
plt.rcParams['ytick.labelsize'] = 11

# Premium visual color palette (Muted Indigo, Steel Blue, Teal, Coral, Gray)
PRIMARY_COLOR = '#4e73df'
SECONDARY_COLOR = '#1cc88a'
ACCENT_COLOR = '#f6c23e'
DANGER_COLOR = '#e74a3b'
DARK_COLOR = '#5a5c69'
palette = [PRIMARY_COLOR, SECONDARY_COLOR, ACCENT_COLOR, DANGER_COLOR, DARK_COLOR]
sns.set_palette(sns.color_palette(palette))

# Set default Plotly template
pio.templates.default = 'plotly_white'


## 💾 5. Load Dataset
We load the NYC Uber rides dataset and display the first few rows to confirm successful ingestion.


In [2]:
import pandas as pd

# Load the dataset from local storage
df = pd.read_csv('final_internship_data.csv')

# Print shape
print(f'Dataset Shape: {df.shape[0]:,} rows, {df.shape[1]} columns')

# Display first 5 rows
df.head()

Dataset Shape: 500,000 rows, 26 columns


,User ID,User Name,Driver Name,Car Condition,Weather,Traffic Condition,key,fare_amount,pickup_datetime,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,passenger_count,hour,day,month,weekday,year,jfk_dist,ewr_dist,lga_dist,sol_dist,nyc_dist,distance,bearing
0,KHVrEVlD,Kimberly Adams,Amy Butler,Very Good,windy,Congested Traffic,2009-06-15 17:26:21.0000001,4.5000,2009-06-15 17:26:21,-1.2888,0.7107,-1.2888,0.7106,1,17,15,6,0,2009,20.2658,55.1760,14.3426,34.5435,27.5726,1.0308,-2.9189
1,lPxIuEri,Justin Tapia,Hannah Zimmerman,Excellent,cloudy,Flow Traffic,2010-01-05 16:52:16.0000002,16.9000,2010-01-05 16:52:16,-1.2918,0.7105,-1.2912,0.7118,1,16,5,1,1,2010,44.6677,31.8324,23.1308,15.1259,8.7557,8.4501,-0.3752
2,gsVN8JLS,Elizabeth Lopez,Amanda Jackson,Bad,stormy,Congested Traffic,2011-08-18 00:35:00.00000049,5.7000,2011-08-18 00:35:00,-1.2912,0.7114,-1.2914,0.7112,2,0,18,8,3,2011,43.5977,33.7121,19.8653,17.7226,9.8473,1.3895,2.6000
3,9I7kWFgd,Steven Wilson,Amy Horn,Very Good,stormy,Flow Traffic,2012-04-21 04:30:42.0000001,7.7000,2012-04-21 04:30:42,-1.2913,0.7109,-1.2914,0.7114,1,4,21,4,5,2012,42.6430,32.5563,21.0631,15.7390,7.7034,2.7993,0.1339
4,8QN5ZaGN,Alexander Andrews,Cassandra Larson,Bad,stormy,Congested Traffic,2010-03-09 07:51:00.000000135,5.3000,2010-03-09 07:51:00,-1.2910,0.7115,-1.2908,0.7118,1,7,9,3,1,2010,43.3300,39.4068,15.2193,23.7324,15.6007,1.9992,-0.5027


In [3]:
# Check schema and data types
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500000 entries, 0 to 499999
Data columns (total 26 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   User ID            500000 non-null  object 
 1   User Name          500000 non-null  object 
 2   Driver Name        500000 non-null  object 
 3   Car Condition      500000 non-null  object 
 4   Weather            500000 non-null  object 
 5   Traffic Condition  500000 non-null  object 
 6   key                500000 non-null  object 
 7   fare_amount        500000 non-null  float64
 8   pickup_datetime    500000 non-null  object 
 9   pickup_longitude   500000 non-null  float64
 10  pickup_latitude    500000 non-null  float64
 11  dropoff_longitude  499995 non-null  float64
 12  dropoff_latitude   499995 non-null  float64
 13  passenger_count    500000 non-null  int64  
 14  hour               500000 non-null  int64  
 15  day                500000 non-null  int64  
 16  mo

In [4]:
# Summary statistics for numerical variables
df.describe().T


,count,mean,std,min,25%,50%,75%,max
fare_amount,500000.0000,11.3584,9.9166,-44.9000,6.0000,8.5000,12.5000,500.0000
pickup_longitude,500000.0000,-1.2657,0.2069,-52.1198,-1.2914,-1.2912,-1.2910,37.3605
pickup_latitude,500000.0000,0.6967,0.1409,-54.3894,0.7110,0.7113,0.7115,29.7246
dropoff_longitude,499995.0000,-1.2658,0.2059,-59.0497,-1.2914,-1.2912,-1.2909,0.7130
dropoff_latitude,499995.0000,0.6967,0.1290,-44.6760,0.7109,0.7113,0.7115,7.0619
passenger_count,500000.0000,1.6834,1.3074,0.0000,1.0000,1.0000,2.0000,6.0000
hour,500000.0000,13.5108,6.5116,0.0000,9.0000,14.0000,19.0000,23.0000
day,500000.0000,15.6842,8.6811,1.0000,8.0000,16.0000,23.0000,31.0000
month,500000.0000,6.2687,3.4378,1.0000,3.0000,6.0000,9.0000,12.0000
weekday,500000.0000,3.0420,1.9492,0.0000,1.0000,3.0000,5.0000,6.0000


In [5]:
# Check for missing values
missing_vals = df.isnull().sum()
missing_percent = 100 * missing_vals / len(df)
missing_df = pd.DataFrame({'Missing Values': missing_vals, 'Percentage (%)': missing_percent})
missing_df[missing_df['Missing Values'] > 0]


,Missing Values,Percentage (%)
dropoff_longitude,5,0.0010
dropoff_latitude,5,0.0010
jfk_dist,5,0.0010
ewr_dist,5,0.0010
lga_dist,5,0.0010
sol_dist,5,0.0010
nyc_dist,5,0.0010
distance,5,0.0010
bearing,5,0.0010


In [6]:
df = df.dropna()

In [7]:
# Check for duplicate rows
duplicates_count = df.duplicated().sum()
print(f'Total duplicate rows in the dataset: {duplicates_count:,}')


Total duplicate rows in the dataset: 0


In [8]:
# Analyze target variable (fare_amount)
print('Target variable summary:')
print(df['fare_amount'].describe())


Target variable summary:
count   499995.0000
mean        11.3582
std          9.9161
min        -44.9000
25%          6.0000
50%          8.5000
75%         12.5000
max        500.0000
Name: fare_amount, dtype: float64


### 📌 Key Findings from Initial Exploration
1. **Missing Values:** There are exactly **5 missing values** in `dropoff_longitude`, `dropoff_latitude`, `jfk_dist`, `ewr_dist`, `lga_dist`, `sol_dist`, `nyc_dist`, `distance`, and `bearing`. Since they affect only 5 rows out of 500,000, we can safely drop them.
2. **Duplicates:** There are **0 duplicate rows** in the dataset.
3. **Anomalies in Target:** The target variable `fare_amount` has a minimum value of **-$44.90** and a maximum of **$500.00**. Negative fares are physically impossible (drivers don't pay passengers), and exceptionally high fares ($500) might represent entry errors or out-of-city long-distance runs that are severe outliers.
4. **Invalid Coordinates:** Min/max coordinates are highly anomalous (e.g. pickup latitude of -54.38 and dropoff longitude of 0.71). NYC is situated around **Latitude 40.7** and **Longitude -74.0**. Values like 0 or numbers outside this region represent corrupted GPS records.
5. **Passenger Count anomalies:** The minimum passenger count is **0**, which is physically invalid for a ride-hailing transaction.
6. **Distance outliers:** The maximum distance is **12,399 km**, which is impossible for a metropolitan ride. These are caused by coordinates that were entered incorrectly.


## 🧹 7. Data Cleaning
Here we execute structured data cleaning steps. For each step, we explain the engineering rationale and execute the code.

### Step 7.1: Dropping Irrelevant Features
**Why it is needed:** The columns `User ID`, `User Name`, `Driver Name`, `key`, and `pickup_datetime` contain unique identifiers, raw strings, and PII. They do not contain any generalized predictive signal.
**Risk of keeping:** Keeping them increases memory usage and causes tree-based models to overfit on noise (especially high-cardinality features like User ID).
**Action:** Drop them.


In [9]:
# Drop irrelevant and high-cardinality columns
columns_to_drop = ['User ID', 'User Name', 'Driver Name', 'key', 'pickup_datetime'  , "ewr_dist", "sol_dist"]

df.drop(columns=columns_to_drop, inplace=True, errors='ignore')
print(f'Columns remaining: {list(df.columns)}')


Columns remaining: ['Car Condition', 'Weather', 'Traffic Condition', 'fare_amount', 'pickup_longitude', 'pickup_latitude', 'dropoff_longitude', 'dropoff_latitude', 'passenger_count', 'hour', 'day', 'month', 'weekday', 'year', 'jfk_dist', 'lga_dist', 'nyc_dist', 'distance', 'bearing']


### Step 7.2: Dropping Missing Values
**Why it is needed:** There are 5 rows with null dropoff coordinates and distances.
**Action:** Drop these rows.


In [10]:
# Drop rows with null values
print(f'Shape after dropping nulls: {df.shape}')


Shape after dropping nulls: (499995, 19)


### Step 7.3: Validating Fare Amount
**Why it is needed:** In NYC, the minimum base taxi fare is $2.50 Negative fares and fares below $2.50 are invalid data.

In [11]:
# Filter fare_amount
df = df[(df['fare_amount'] >= 2.5) ]
print(f'Shape after fare filtering: {df.shape}')


Shape after fare filtering: (499954, 19)


### Step 7.4: Validating Coordinates (NYC Bounding Box)
**Why it is needed:** NYC taxi rides must start and end within metropolitan limits. Coordinates outside this zone (e.g. 0,0) represent GPS glitches.
**Bounding Box limits (in degrees):**
- Longitude: $[-74.30, -73.65]$
- Latitude: $[40.45, 40.95]$
**Note on Radians:** The coordinates in the dataset are already stored in **radians**. Thus, we must convert our bounding box boundaries to radians before filtering.
**Action:** Apply the bounding box filter.


In [12]:
# Define NYC bounding box boundaries in degrees and convert to radians
lon_min, lon_max = np.deg2rad(-74.30), np.deg2rad(-73.65)
lat_min, lat_max = np.deg2rad(40.45), np.deg2rad(40.95)

# Apply bounding box filter
df = df[
    df['pickup_longitude'].between(lon_min, lon_max) &
    df['pickup_latitude'].between(lat_min, lat_max) &
    df['dropoff_longitude'].between(lon_min, lon_max) &
    df['dropoff_latitude'].between(lat_min, lat_max)
]
print(f'Shape after coordinate filtering: {df.shape}')

Shape after coordinate filtering: (489015, 19)


### Step 7.5: Validating Passenger Count
**Why it is needed:** Passenger count must be between 1 and 4. Rides with 0 passengers are invalid. Passenger counts of 5 or 6 (UberXL) represent a separate pricing tier. To build a standard, reliable prediction model, we restrict passenger count to $[1, 4]$ and note that UberXL/SUV trips should be modeled separately.
**Action:** Restrict passenger_count to $[1, 4]$.


In [13]:
# Filter passenger_count
df = df[(df['passenger_count'] >= 1) & (df['passenger_count'] <= 4)]
print(f'Shape after passenger count filtering: {df.shape}')


Shape after passenger count filtering: (442322, 19)


### Step 7.6: Validating Ride Distance
**Why it is needed:** Distance values above 100 km are extreme outliers for NYC metropolitan taxi trips. We also filter out 0 distance trips since they represent cancellations, waits, or GPS errors, rather than actual travel distance.
**Action:** Filter distance to $(0, 100]$ km.


In [14]:
# Filter distance
logical_max_distance = 100.0
df = df[(df['distance'] > 0.0) & (df['distance'] <= logical_max_distance)]
print(f'Final cleaned dataset shape: {df.shape}')


Final cleaned dataset shape: (437739, 19)


## ⚙️ 8. Feature Engineering
Feature engineering is crucial for enabling algorithms (especially linear regression and neural networks) to learn complex patterns. The dataset already has calculated great-circle distance and airport distances, but we will add **Cyclic Temporal Encodings**.

### Cyclic Temporal Encoding
Temporal features like hour of the day (0-23), day of the week (0-6), and month of the year (1-12) are circular. For example, 23:00 (11 PM) is as close to 00:00 (Midnight) as 01:00 is to 02:00. However, in standard numerical representation, the distance between 23 and 0 is 23. This confuses distance-based models.

To resolve this, we map these features onto a 2D circle using sine and cosine transformations:
$$x_{sin} = \sin\left(\frac{2\pi \cdot x}{\text{Period}}\right)$$
$$x_{cos} = \cos\left(\frac{2\pi \cdot x}{\text{Period}}\right)$$

#### **Advantages:**
- Maps circular time sequences continuously, preserving proximity relationships (e.g. 23:00 close to 0:00).
- Crucial for Neural Networks, Support Vector Regressors, and Linear/Ridge Regression models.

#### **Disadvantages:**
- Adds dimensionality (doubles the number of temporal columns).
- Tree-based models (XGBoost, Random Forests) don't naturally benefit from this because they partition variables using splits, though it does not harm their performance.

Let's execute cyclic encoding for **hour**, **weekday**, and **month**.


In [15]:
# Cyclic encoding for hour (period = 24)
df['hour_sin'] = np.sin(2 * np.pi * df['hour'] / 24.0)
df['hour_cos'] = np.cos(2 * np.pi * df['hour'] / 24.0)

# Cyclic encoding for weekday (period = 7)
df['weekday_sin'] = np.sin(2 * np.pi * df['weekday'] / 7.0)
df['weekday_cos'] = np.cos(2 * np.pi * df['weekday'] / 7.0)

# Cyclic encoding for month (period = 12)
df['month_sin'] = np.sin(2 * np.pi * (df['month'] - 1) / 12.0)
df['month_cos'] = np.cos(2 * np.pi * (df['month'] - 1) / 12.0)

# Display engineered features
df[['hour', 'hour_sin', 'hour_cos', 'weekday', 'weekday_sin', 'weekday_cos']].head()


,hour,hour_sin,hour_cos,weekday,weekday_sin,weekday_cos
0,17,-0.9659,-0.2588,0,0.0000,1.0000
1,16,-0.8660,-0.5000,1,0.7818,0.6235
2,0,0.0000,1.0000,3,0.4339,-0.9010
3,4,0.8660,0.5000,5,-0.9749,-0.2225
4,7,0.9659,-0.2588,1,0.7818,0.6235


In [16]:
df["is_weekend"] = df["weekday"].isin([5, 6]).astype("int8")

df["rush_hour"] = (
    ((df["hour"] >= 7) & (df["hour"] <= 9)) |
    ((df["hour"] >= 17) & (df["hour"] <= 22))
).astype("int8")

df["is_night"] = (
    (df["hour"] >= 23) | (df["hour"] <= 5)
).astype("int8")

In [17]:
col_drop = ['hour' , 'day' , 'month' , 'weekday']

df.drop(columns=col_drop, inplace=True)

In [18]:
df.head(8)

,Car Condition,Weather,Traffic Condition,fare_amount,pickup_longitude,pickup_latitude,dropoff_longitude,dropoff_latitude,passenger_count,year,jfk_dist,lga_dist,nyc_dist,distance,bearing,hour_sin,hour_cos,weekday_sin,weekday_cos,month_sin,month_cos,is_weekend,rush_hour,is_night
0,Very Good,windy,Congested Traffic,4.5000,-1.2888,0.7107,-1.2888,0.7106,1,2009,20.2658,14.3426,27.5726,1.0308,-2.9189,-0.9659,-0.2588,0.0000,1.0000,0.5000,-0.8660,0,1,0
1,Excellent,cloudy,Flow Traffic,16.9000,-1.2918,0.7105,-1.2912,0.7118,1,2010,44.6677,23.1308,8.7557,8.4501,-0.3752,-0.8660,-0.5000,0.7818,0.6235,0.0000,1.0000,0,0,0
2,Bad,stormy,Congested Traffic,5.7000,-1.2912,0.7114,-1.2914,0.7112,2,2011,43.5977,19.8653,9.8473,1.3895,2.6000,0.0000,1.0000,0.4339,-0.9010,-0.5000,-0.8660,0,0,1
3,Very Good,stormy,Flow Traffic,7.7000,-1.2913,0.7109,-1.2914,0.7114,1,2012,42.6430,21.0631,7.7034,2.7993,0.1339,0.8660,0.5000,-0.9749,-0.2225,1.0000,0.0000,1,0,1
4,Bad,stormy,Congested Traffic,5.3000,-1.2910,0.7115,-1.2908,0.7118,1,2010,43.3300,15.2193,15.6007,1.9992,-0.5027,0.9659,-0.2588,0.7818,0.6235,0.8660,0.5000,0,1,0
5,Excellent,sunny,Congested Traffic,12.1000,-1.2916,0.7109,-1.2911,0.7114,1,2011,42.3356,20.6482,7.6495,3.7872,-0.6743,0.7071,-0.7071,0.4339,-0.9010,0.0000,1.0000,0,1,0
6,Excellent,windy,Flow Traffic,7.5000,-1.2912,0.7113,-1.2911,0.7115,1,2012,42.5632,18.1137,10.9931,1.5558,-0.3423,-0.8660,0.5000,0.7818,0.6235,-0.8660,0.5000,0,1,0
7,Good,sunny,Congested Traffic,16.5000,-1.2907,0.7116,-1.2914,0.7112,1,2012,42.5332,16.9495,12.4492,4.1554,2.2366,-0.9659,-0.2588,0.9749,-0.2225,0.0000,1.0000,0,1,0


In [19]:
df["Car Condition"] = df["Car Condition"].map({
    "Bad": 0,
    "Good": 1,
    "Very Good": 2,
    "Excellent": 3
})

df["Traffic Condition"] = df["Traffic Condition"].map({
    "Flow Traffic": 0,
    "Dense Traffic": 1,
    "Congested Traffic": 2
})

In [20]:
x = df.drop('fare_amount' , axis= 1)
y = (df["fare_amount"])

In [21]:
from sklearn.model_selection import train_test_split

x_train , x_test  , y_train  , y_test = train_test_split( x , y , test_size= 0.2 , random_state= 5 )

In [22]:
numeric_features = [
    "Car Condition",
    "Traffic Condition",
    "pickup_longitude",
    "pickup_latitude",
    "dropoff_longitude",
    "dropoff_latitude",
    "passenger_count",
    "year",
    "jfk_dist",

    "lga_dist",
    "nyc_dist",
    "distance",
    "bearing",
    "hour_sin",
    "hour_cos",
    "weekday_sin",
    "weekday_cos",
    "month_sin",
    "month_cos",
    "rush_hour",
    "is_weekend",
    "is_night",

]

categorical_features = [
    "Weather"
]

In [23]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import RobustScaler, OneHotEncoder

preprocessor = ColumnTransformer(
    transformers=[
        ("num", RobustScaler(), numeric_features),
        ("weather", OneHotEncoder(handle_unknown="ignore"), categorical_features)
    ]
)

In [24]:
!pip install catboost


In [24]:
from catboost import CatBoostRegressor

model = CatBoostRegressor(
    random_state=42,
    verbose=0,
    loss_function="RMSE",
    task_type="GPU",
    devices="0",
    bootstrap_type="Bernoulli"
)

In [25]:
from sklearn.pipeline import Pipeline

pipeline = Pipeline([
    ("preprocessor", preprocessor),
    ("model", model)
])

In [26]:
from sklearn.model_selection import RandomizedSearchCV

param_dist = {

    "model__depth": [4,6,8,10],

    "model__learning_rate": [0.03,0.05,0.1],

    "model__iterations": [300,500,700],

    "model__l2_leaf_reg": [1,3,5,7,9],

    "model__subsample": [0.7,0.8,0.9,1.0]

}

random_search = RandomizedSearchCV(

    estimator=pipeline,

    param_distributions=param_dist,

    n_iter=10,

    scoring="neg_root_mean_squared_error",

    cv=5,

    random_state=42,

    n_jobs=1,

    verbose=2

)

random_search.fit(x_train, y_train)

print("Best Parameters:")
print(random_search.best_params_)

print("\nBest CV RMSE:")
print(-random_search.best_score_)

Fitting 5 folds for each of 10 candidates, totalling 50 fits
[CV] END model__depth=6, model__iterations=700, model__l2_leaf_reg=7, model__learning_rate=0.05, model__subsample=0.7; total time=   4.2s
[CV] END model__depth=6, model__iterations=700, model__l2_leaf_reg=7, model__learning_rate=0.05, model__subsample=0.7; total time=   3.5s
[CV] END model__depth=6, model__iterations=700, model__l2_leaf_reg=7, model__learning_rate=0.05, model__subsample=0.7; total time=   3.5s
[CV] END model__depth=6, model__iterations=700, model__l2_leaf_reg=7, model__learning_rate=0.05, model__subsample=0.7; total time=   3.7s
[CV] END model__depth=6, model__iterations=700, model__l2_leaf_reg=7, model__learning_rate=0.05, model__subsample=0.7; total time=   3.7s
[CV] END model__depth=6, model__iterations=500, model__l2_leaf_reg=9, model__learning_rate=0.03, model__subsample=0.9; total time=   2.8s
[CV] END model__depth=6, model__iterations=500, model__l2_leaf_reg=9, model__learning_rate=0.03, model__subsamp

In [27]:
best_model = random_search.best_estimator_

In [28]:
pred = best_model.predict(x_test)

In [29]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score
import numpy as np

pred = best_model.predict(x_test)

print("MAE :", mean_absolute_error(y_test, pred))
print("RMSE:", np.sqrt(mean_squared_error(y_test, pred)))
print("R2 :", r2_score(y_test, pred))

MAE : 1.5823869855404313
RMSE: 3.3470432887388486
R2 : 0.8735654576223247


In [30]:
results = pd.DataFrame({
    "Actual": y_test,
    "Predicted": pred
})

results["Error"] = abs(
    results["Actual"] - results["Predicted"]
)

In [31]:
results.head()

,Actual,Predicted,Error
219791,4.5000,5.0331,0.5331
406385,5.3000,4.6096,0.6904
445360,5.3000,5.2353,0.0647
323812,9.0000,10.6139,1.6139
51875,6.5000,6.9597,0.4597


In [ ]:
from pathlib import Path

MODEL_PATH = Path("models") / "catboost_fare_pipeline.joblib"
model = joblib.load(MODEL_PATH)

['uber_fare_preprocessor.joblib']

In [33]:
print(best_model)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num', RobustScaler(),
                                                  ['Car Condition',
                                                   'Traffic Condition',
                                                   'pickup_longitude',
                                                   'pickup_latitude',
                                                   'dropoff_longitude',
                                                   'dropoff_latitude',
                                                   'passenger_count', 'year',
                                                   'jfk_dist', 'lga_dist',
                                                   'nyc_dist', 'distance',
                                                   'bearing', 'hour_sin',
                                                   'hour_cos', 'weekday_sin',
                                                   'weekday_cos', 'month_sin',
           

## Model Development Summary

During this project, multiple regression models and preprocessing strategies were evaluated to improve fare prediction performance.

| Experiment | Description | MAE | RMSE | R² | Notes |
|:----------|:------------|----:|-----:|---:|:------|
| Baseline Linear Regression | Initial model before advanced tuning | 2.1596 | 4.2184 | 0.7992 | Baseline benchmark |
| Linear Regression (CV) |  | 2.1790 | 4.2817 | 0.7931 | Similar performance to Ridge |
| Ridge Regression | L2 Regularization | 2.1790 | 4.2817 | 0.7931 | No noticeable improvement |
| Lasso Regression | L1 Regularization | 2.4208 | 4.5785 | 0.7634 | Lowest performance |
| Random Forest (Tuned) | RandomizedSearchCV | **1.5971** | **3.3126** | **0.8762** | Strong overall performance |
| XGBoost (Tuned) | RandomizedSearchCV | **≈1.54** | **≈3.31** | **≈0.87** | Lowest MAE among boosted models |
| CatBoost (Tuned) | RandomizedSearchCV | **1.5778** | **3.3117** | **0.8762** | Best balance between RMSE and R² |
| CatBoost + Log(Target) | Training using log1p(target) | **1.4830** | 3.3698 | 0.8718 | Lower MAE, but RMSE and R² degraded |
| Outlier Removal Analysis | Comparing dataset before/after filtering | 1.5555 | **2.2820*** | 0.6992 | RMSE improved significantly, but R² decreased because target variance changed after filtering. Metrics are **not directly comparable** to the original dataset. |

> *The RMSE after outlier removal was computed on a different target distribution; therefore, it should not be directly compared with the previous experiments.*

### Key Findings

- Linear Regression served as the baseline model with an RMSE of **4.22**.
- Ridge Regression produced nearly identical results to Linear Regression.
- Lasso Regression underperformed due to stronger regularization.
- Random Forest substantially improved prediction accuracy after hyperparameter tuning.
- XGBoost achieved the lowest MAE among the gradient boosting models.
- CatBoost achieved the best overall trade-off between RMSE and R² while requiring minimal categorical preprocessing.
- Applying a logarithmic transformation to the target reduced MAE but slightly worsened RMSE and R² after inverse transformation.
- Outlier removal greatly reduced prediction error; however, the resulting metrics cannot be fairly compared with previous experiments because they were evaluated on a different target distribution.

## Final Conclusion

Several regression algorithms were explored throughout this project, including Linear Regression, Ridge, Lasso, Random Forest, XGBoost, and CatBoost.

After extensive experimentation with feature engineering, preprocessing, cross-validation, and hyperparameter optimization, **CatBoost** and **Random Forest** consistently delivered the strongest overall performance, achieving an RMSE of approximately **3.31** and an R² of **0.876**.

Although XGBoost achieved a slightly lower MAE, CatBoost demonstrated a more balanced performance across all evaluation metrics while integrating seamlessly into the preprocessing pipeline.

Overall, CatBoost was selected as the final model due to its robust predictive performance, stability during cross-validation, and strong generalization on the unseen test set.